File: backend/app/schemas/recommendation.py
from typing import Literal, Optional

from pydantic import BaseModel, Field, field_validator

from app.schemas.job import JobPostingRead


RoleCategory = Literal[
    "backend",
    "frontend",
    "fullstack",
    "infra",
    "ai",
    "data",
    "qa",
    "security",
    "embedded",
    "mobile",
    "enterprise",
]


class UserProfileInput(BaseModel):
    skills: list[str] = Field(min_length=1, examples=[["Python", "Linux", "SQL"]])
    experience_years: int = Field(ge=0, le=50, examples=[1])
    japanese_level: Optional[Literal["N1", "N2", "N3"]] = None
    desired_role: Optional[RoleCategory] = None
    preferred_locations: Optional[list[str]] = Field(default=None, examples=[["东京", "大阪"]])
    expected_salary: Optional[int] = Field(default=None, ge=0, examples=[300000])
    project_summary: Optional[str] = Field(default=None, max_length=2000)

    @field_validator("skills", "preferred_locations")
    @classmethod
    def remove_empty_values(cls, values: Optional[list[str]]) -> Optional[list[str]]:
        if values is None:
            return None
        cleaned = [value.strip() for value in values if value.strip()]
        if not cleaned:
            raise ValueError("列表中至少需要一个有效值")
        return cleaned


class ScoreBreakdown(BaseModel):
    required_skills: float
    preferred_skills: float
    experience: float
    japanese: float
    role: float
    location: float
    salary: float


class RecommendationItem(BaseModel):
    job: JobPostingRead
    match_score: float
    score_breakdown: ScoreBreakdown
    reasons: list[str]
    missing_skills: list[str]
    next_steps: list[str]


class RecommendationResponse(BaseModel):
    total_jobs_evaluated: int
    recommendations: list[RecommendationItem]

File: frontend/src/types/recommendation.ts
export type JapaneseLevel = "N1" | "N2" | "N3";

export type RoleCategory =
  | "backend"
  | "frontend"
  | "fullstack"
  | "infra"
  | "ai"
  | "data"
  | "qa"
  | "security"
  | "embedded"
  | "mobile"
  | "enterprise";

// 这些类型与 FastAPI 的 Pydantic Schema 一一对应，字段写错时 TypeScript 会提前提醒。
export interface UserProfileInput {
  skills: string[];
  experience_years: number;
  japanese_level?: JapaneseLevel;
  desired_role?: RoleCategory;
  preferred_locations?: string[];
  expected_salary?: number;
  project_summary?: string;
}

export interface JobPosting {
  id: number;
  title: string;
  role_category: string;
  company_name: string;
  company_type: string;
  location: string;
  industry: string;
  required_skills: string[];
  preferred_skills: string[] | null;
  japanese_level: string | null;
  experience_requirement: string;
  initial_salary: number;
  job_url: string;
  created_at: string;
}

export interface ScoreBreakdown {
  required_skills: number;
  preferred_skills: number;
  experience: number;
  japanese: number;
  role: number;
  location: number;
  salary: number;
}

export interface RecommendationItem {
  job: JobPosting;
  match_score: number;
  score_breakdown: ScoreBreakdown;
  reasons: string[];
  missing_skills: string[];
  next_steps: string[];
}

export interface RecommendationResponse {
  total_jobs_evaluated: number;
  recommendations: RecommendationItem[];
}
